earth_enterprise/src/common/SharedString.h

```diff
@@ -42,30 +43,42 @@ class SharedString {
         std::unordered_map<uint32_t, std::string> refFromKeyTable;
         std::unordered_map<std::string, uint32_t> keyFromRefTable;
         static uint32_t nextID;
+        mutable pthread_rwlock_t rwlock;
+        mutable int rc = pthread_rwlock_init(&rwlock, NULL);
       public:
         const std::string & RefFromKey(const uint32_t &key) {
+          rc = pthread_rwlock_rdlock(&rwlock);
```

Comment:
Locking and unlocking is typically done in some kind of wrapper class. The reasoning is that if something bad happens (an exception is thrown in the middle of the function) the lock will still be released. There's an example of how to do this in the parallel-system-manager branch: https://github.com/google/earthenterprise/blob/parallel-system-manager/earth_enterprise/src/common/khMTTypes.h (lines 125-160). To acquire a read lock you just create an instance of the khReadGuard, passing a khRWLock as a parameter. The lock is released when the read guard goes out of scope. I'll leave it up to you as to whether you want to make that change here.